File: leetcode/0074-search-a-2d-matrix/solution.py
class Solution(object):
    def searchMatrix(self, matrix, target):
        m = len(matrix)
        n = len(matrix[0])

        # Binary search over rows to find the correct row
        top = 0
        bottom = m - 1
        target_row = -1

        while top <= bottom:
            mid = (top + bottom) // 2
            row = matrix[mid]
            if row[0] <= target <= row[-1]:
                target_row = mid
                break
            elif target < row[0]:
                bottom = mid - 1
            else:
                top = mid + 1

        if target_row == -1:
            return False

        # Binary search within the identified row
        row = matrix[target_row]
        left = 0
        right = n - 1

        while left <= right:
            mid = (left + right) // 2
            if row[mid] == target:
                return True
            elif row[mid] < target:
                left = mid + 1
            else:
                right = mid - 1

        return False

File: leetcode/0074-search-a-2d-matrix/solution.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m=matrix.size();
        int n=matrix[0].size();
        int left=0;
        int right=m-1;
        int r=-1;
        while(left<=right){
            int mid=(right-left)/2+left;
            if(matrix[mid][0]<=target && matrix[mid][n-1]>=target){
                r=mid;
                break;
            }
            else if(target>matrix[mid][0]){
                left=mid+1;
            }
            else{
                right=mid-1;
            }
        }
        if(r==-1){
            return false;
        }
        vector<int> row=matrix[r];
        left=0;
        right=n-1;
        while(left<=right){
            int mid=(right-left)/2+left;
            if(row[mid]==target){
                return true;
            }
            else if(row[mid]<target){
                left=mid+1;
            }
            else{
                right=mid-1;
            }
        }
        return false;
    }
};
